R/submit-models.R

```diff
@@ -128,6 +128,11 @@ submit_nonmem_models <- function(.mods,
         cmd_args,
         sprintf("--config=%s", normalizePath(.config_path))
       )
+    } else {
+      default_config_path <- file.path(model_dir, "bbi.yaml")
+      if (!fs::file_exists(default_config_path) && !isTRUE(.dry_run)) {
```

Comment:
I'd say that the closer a dry-run mode behaves to the regular mode the better, so it'd be good if it would include this "file exists" guard.  Three tests fail with that change, but my initial impression is that that is more of an issue with those tests than with enabling the guard for dry runs.

Perhaps I'm missing something.  Can you explain more why this check should be bypassed when `.dry_run = TRUE`, @barrettk (or @seth127)?  I see `.dry_run` mentioned a bit at gh-495, but nothing that seems to lay that out.

Anyway, here's what that direction could look like, including the necessary test tweaks:

<details>
<summary>diff</summary>

```diff
diff --git a/R/submit-models.R b/R/submit-models.R
index 19c9ca8..8b79729 100644
--- a/R/submit-models.R
+++ b/R/submit-models.R
@@ -84,7 +84,7 @@ submit_models.list <- function(
 #' @importFrom stringr str_detect
 #' @importFrom tools file_path_sans_ext
 #' @importFrom purrr map
-#' @importFrom rlang is_bare_list
+#' @importFrom rlang %||% is_bare_list
 #' @return A list of S3 objects of class `bbi_process`
 #' @keywords internal
 submit_nonmem_models <- function(.mods,
@@ -122,17 +122,14 @@ submit_nonmem_models <- function(.mods,
 
     model_dir <- get_model_working_directory(.run[["models"]][[1L]])
 
+    checkmate::assert_file_exists(
+      .config_path %||% file.path(model_dir, "bbi.yaml"))
+
     if (!is.null(.config_path)) {
-      checkmate::assert_file_exists(.config_path)
       cmd_args <- c(
         cmd_args,
         sprintf("--config=%s", normalizePath(.config_path))
       )
-    } else {
-      default_config_path <- file.path(model_dir, "bbi.yaml")
-      if (!fs::file_exists(default_config_path) && !isTRUE(.dry_run)) {
-        stop(glue("Looking for default configuration file at {default_config_path}: no such file or directory"), call. = F)
-      }
     }
 
     return(list(cmd_args = cmd_args, model_dir = model_dir))
diff --git a/tests/testthat/test-print.R b/tests/testthat/test-print.R
index 8763295..275fb18 100644
--- a/tests/testthat/test-print.R
+++ b/tests/testthat/test-print.R
@@ -37,6 +37,11 @@
     temp_dir <- file.path(get_model_working_directory(MOD1), "print-test")
     fs::dir_create(temp_dir)
     on.exit(fs::dir_delete(temp_dir))
+
+    withr::with_options(list(bbr.bbi_exe_path = read_bbi_path()), {
+      bbi_init(temp_dir, ".", .no_default_version = TRUE)
+    })
+
     mods <- purrr::map(1:50, ~copy_model_from(MOD1, file.path("print-test", .x)))
     proc <- submit_models(mods, .dry_run = TRUE)
 
diff --git a/tests/testthat/test-submit-models.R b/tests/testthat/test-submit-models.R
index b1c135e..02e444b 100644
--- a/tests/testthat/test-submit-models.R
+++ b/tests/testthat/test-submit-models.R
@@ -72,10 +72,14 @@
             })
 
   test_that("submit_models() works for models in different directories [BBR-SBMT-011]", {
-    new_dir <- "level2"
-    fs::dir_create(file.path(MODEL_DIR, new_dir))
+    new_dir <- file.path(ABS_MODEL_DIR, "level2")
+    fs::dir_create(new_dir)
     on.exit(cleanup())
 
+    withr::with_options(list(bbr.bbi_exe_path = read_bbi_path()), {
+      bbi_init(new_dir, ".", .no_default_version = TRUE)
+    })
+
     # TODO: use test helper functions, e.g., create_all_models(), once the
     # model_directory option is deprecated
     mod2 <- copy_model_from(
@@ -160,6 +164,10 @@
     mod <- read_model(temp_mod_path)
     mod <- replace_all_bbi_args(mod, NULL)
 
+    withr::with_options(list(bbr.bbi_exe_path = read_bbi_path()), {
+      bbi_init(dirname(temp_mod_path), ".", .no_default_version = TRUE)
+    })
+
     res <- submit_models(list(mod), .dry_run = TRUE)
 
     expect_identical(
```

</details>

Reply:
@seth127
>  I guess its debatable whether being able to run .dry_run=T without a valid config is actually "desired behavior" (from a UX standpoint) or not

Yes, I think so.  I lean toward a dry run erroring for this (similar to how a dry-run `git push` would error if, say, you didn't have permissions to push to the remote).  But I'm okay with keeping the current `submit_model()` behavior as long as the case is being made that it is better from UX perspective or that it's needed for backward compatibility, as opposed to it being something done for testing convenience.

A related point: It seems inconsistent that a dry run would fail when a config is explicitly given but not for the default.  Perhaps that is done for the sake of the `normalizePath` call, but that doesn't change that it's inconsistent from the UI perspective.

> I will say that I think we should keep submit_model() and submit_models() consistent

Right, sorry, I didn't step back from the changes of this PR when creating that patch, but I very much agree with this.

@barrettk 
> I'll test out Kyle's solution and see If I can tweak it all

Thanks.